jupyter-extension/src/plugins/mount/components/Datum/hooks/useDatum.tsx

```diff
@@ -0,0 +1,130 @@
+import {requestAPI} from '../../../../../handler';
+import {useEffect, useState} from 'react';
+import {ServerConnection} from '@jupyterlab/services';
+import {DatumsResponse} from 'plugins/mount/types';
+
+export type useDatumResponse = {
+  loading: boolean;
+  shouldShowCycler: boolean;
+  currentDatumId: string;
+  currentDatumIdx: number;
+  setCurrentDatumIdx: (idx: number) => void;
+  numDatums: number;
+  inputSpec: string;
+  setInputSpec: (input: string) => void;
+  callMountDatums: () => Promise<void>;
+  callUnmountAll: () => Promise<void>;
+  errorMessage: string;
+};
+
+export const useDatum = (
+  showDatum: boolean,
+  keepMounted: boolean,
+  refresh: () => void,
+  pollRefresh: () => Promise<void>,
+  currentDatumInfo?: DatumsResponse,
+): useDatumResponse => {
+  const [loading, setLoading] = useState(false);
+  const [shouldShowCycler, setShouldShowCycler] = useState(false);
+  const [currentDatumId, setCurrentDatumId] = useState('');
+  const [currentDatumIdx, setCurrentDatumIdx] = useState(-1);
+  const [numDatums, setNumDatums] = useState(-1);
+  const [inputSpec, setInputSpec] = useState('');
+  const [errorMessage, setErrorMessage] = useState('');
+
+  useEffect(() => {
+    if (showDatum && currentDatumIdx !== -1) {
+      callShowDatum();
+    }
+  }, [currentDatumIdx, showDatum]);
+
+  useEffect(() => {
+    if (showDatum && !keepMounted) {
+      callUnmountAll();
+    }
+    if (keepMounted && currentDatumInfo) {
+      setShouldShowCycler(true);
+      setCurrentDatumIdx(currentDatumInfo.curr_idx);
+      setNumDatums(currentDatumInfo.num_datums);
+      setInputSpec(JSON.stringify(currentDatumInfo.input, null, 2));
+    }
+  }, [showDatum]);
+
+  const callMountDatums = async () => {
+    setLoading(true);
+    setErrorMessage('');
+
+    try {
+      const res = await requestAPI<any>('_mount_datums', 'PUT', {
+        input: JSON.parse(inputSpec),
+      });
+      refresh();
+      setCurrentDatumId(res.id);
+      setCurrentDatumIdx(res.idx);
+      setNumDatums(res.num_datums);
```

Comment:
Instead of keeping a state object for each of the properties you could just have one piece of state that was of type `DatumsResponse`

Reply:
Bundled up the items into a `MountDatumsResponse` object. I use this object to refer to data for displaying in the component. I'm using a separate state variable `currIdx`, which keeps track of which datum in the list we've mounted, that triggers the request to mount the next datum upon modification.